### PR TITLE
calculate subnet range if used shared vpc with existed subnets.

### LINKED ISSUE #2294: Error creating subnet The CIDR 'xxx.xxx.xxx.xxx/xx' conflicts with another subnet
When creating a cluster inside an existing VPC which contains already some subnets,
we get errors like:
```
W0405 11:59:18.655454   35343 executor.go:109] error running task "Subnet/utility-eu-west-1a.k8s.eu-west-1.test.redacted.net" (9s remaining to succeed): error creating subnet: InvalidSubnet.Conflict: The CIDR '172.31.0.0/22' conflicts with another subnet
	status code: 400, request id: 9d2fccda-c45b-422d-8ad4-b18add7f4ba4
```

## Expected

Kops should find some non-conflicting IP ranges for the subnets it creates.

## Impact

When demo-ing Kops "Look guys, this tool is awesome, you get a resilient cluster in your AWS account with just 2 commands!"
--> errors happen requiring complicated looking workarounds (see below)
Developers leave thinking "It really doesn't seem so easy".

## Set up
In our existing VPC, we have 3 subnets, with IP ranges:
172.31.0.0/20
172.31.16.0/20
172.31.32.0/20

```
$ kops create cluster \
  --cloud aws \
  --vpc vpc-redacted \
  --master-zones eu-west-1a,eu-west-1b,eu-west-1c \
  --zones eu-west-1a,eu-west-1b,eu-west-1c \
  --name k8s.eu-west-1.test.redacted.net \
  --master-size t2.medium \
  --node-size t2.medium \
  --cloud-labels "Owner=Me,Stack=K8s-test" \
  --networking calico \
  --target cloudformation \
  --state s3://k8s \
  --dns-zone=test.redacted.net \
  --topology private

$ kops update cluster --state s3://k8s --name k8s.eu-west-1.test.redacted.net --yes
```

---> errors happen:
```
W0405 11:59:18.655454   35343 executor.go:109] error running task "Subnet/utility-eu-west-1a.k8s.eu-west-1.test.redacted.net" (9s remaining to succeed): error creating subnet: InvalidSubnet.Conflict: The CIDR '172.31.0.0/22' conflicts with another subnet
	status code: 400, request id: 9d2fccda-c45b-422d-8ad4-b18add7f4ba4
W0405 11:59:18.655497   35343 executor.go:109] error running task "Subnet/utility-eu-west-1b.k8s.eu-west-1.test.redacted.net" (9s remaining to succeed): error creating subnet: InvalidSubnet.Conflict: The CIDR '172.31.4.0/22' conflicts with another subnet
	status code: 400, request id: 4733e80b-46c3-4d53-83fb-6e9ab6cbeb0b
W0405 11:59:18.655514   35343 executor.go:109] error running task "Subnet/utility-eu-west-1c.k8s.eu-west-1.test.redacted.net" (9s remaining to succeed): error creating subnet: InvalidSubnet.Conflict: The CIDR '172.31.8.0/22' conflicts with another subnet
	status code: 400, request id: 75f69b34-7a1a-4ba0-9cac-8fd07fb0e4c8
W0405 11:59:18.655529   35343 executor.go:109] error running task "Subnet/eu-west-1a.k8s.test.eu-west-1.testqbi.stylight.net" (9s remaining to succeed): error creating subnet: InvalidSubnet.Conflict: The CIDR '172.31.32.0/19' conflicts with another subnet
	status code: 400, request id: 463ec4be-3728-4e16-948d-ffb289f72e7d
I0405 11:59:18.655590   35343 executor.go:124] No progress made, sleeping before retrying 4 failed task(s)
```

Our existing subnets in the VPC...
172.31.0.0/20 (from 172.31.0.1 to 172.31.15.254)
172.31.16.0/20 (from 172.31.16.1 to 172.31.31.254)
172.31.32.0/20 (from 172.31.32.1 to 172.31.47.254)

...do indeed conflict with the ones Kops is trying to create:
172.31.0.0/22  (from 172.31.0.1 to 172.31.3.254)   -> conflicts with Subnet with range 172.31.0.0/20!
172.31.4.0/22  (from 172.31.4.1 to 172.31.7.254)   -> conflicts with Subnet with range 172.31.0.0/20!
172.31.8.0/22  (from 172.31.8.1 to 172.31.11.254)  -> conflicts with Subnet with range 172.31.0.0/20!
172.31.32.0/19 (from 172.31.32.1 to 172.31.63.254) -> conflicts with Subnet with range 172.31.32.0/20!

## Workaround
Export the config, fix the subnets and replace the config.
```
$ kops get clusters k8s.eu-west-1.test.redacted.net --state s3://k8s -o yaml > cluster-config.yaml
$ vi  cluster-config.yaml

# in this file, in the Subnet section, replace conflicting IP ranges...
  subnets:
  - cidr: 172.31.32.0/19
    name: eu-west-1a
    type: Private
    zone: eu-west-1a
  - cidr: 172.31.64.0/19
    name: eu-west-1b
    type: Private
    zone: eu-west-1b
  - cidr: 172.31.96.0/19
    name: eu-west-1c
    type: Private
    zone: eu-west-1c
  - cidr: 172.31.0.0/22
    name: utility-eu-west-1a
    type: Utility
    zone: eu-west-1a
  - cidr: 172.31.4.0/22
    name: utility-eu-west-1b
    type: Utility
    zone: eu-west-1b
  - cidr: 172.31.8.0/22
    name: utility-eu-west-1c
    type: Utility
    zone: eu-west-1c

#... by non-conflicting IP ranges, ex:

subnets:
  - cidr: 172.31.224.0/19
    name: eu-west-1a
    type: Private
    zone: eu-west-1a
  - cidr: 172.31.64.0/19
    name: eu-west-1b
    type: Private
    zone: eu-west-1b
  - cidr: 172.31.96.0/19
    name: eu-west-1c
    type: Private
    zone: eu-west-1c
  - cidr: 172.31.128.0/22
    name: utility-eu-west-1a
    type: Utility
    zone: eu-west-1a
  - cidr: 172.31.160.0/22
    name: utility-eu-west-1b
    type: Utility
    zone: eu-west-1b
  - cidr: 172.31.192.0/22
    name: utility-eu-west-1c
    type: Utility
    zone: eu-west-1c


$ kops replace -f cluster-config.yaml --state s3://k8s

$ kops update cluster --state s3://k8s --name k8s.eu-west-1.test.redacted.stylight.net --yes
```
--> then it works

## Version

Kops 1.5.3

### PR DESCRIPTION
**Pull request by pronix**
There was error during creation k8s cluster in existed vpc with subnets.
this patch calculated biggest free subnet in vpc and use it for calculate k8s subnets.
close #2294 
https://github.com/kubernetes/kops/issues/2394

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2395)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Did you mean to add the weave code in your PR?

**Comment by chrislovecnm** [comment]
@justinsb would we want weave vendoring in a different PR?

**Comment by pronix** [comment]
@chrislovecnm without vendoring tests are broken.

**Comment by chrislovecnm** [comment]
@pronix completely understand, about the tests.  What I was asking @justinsb is if we wanted to bring in the vendoring first, and your code second.  It is just a preference of mine, to have separate PRs.  So many filesssssss :)

File: upup/pkg/fi/cloud.go [hunk 1]
```diff
@@ -169,3 +173,41 @@ func GuessCloudForZone(zone string) (CloudProviderID, bool) {
...
+func (s AddrSlice) Less(i, j int) bool { return s[i] < s[j] }
+func (s AddrSlice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// input vpc info, output biggest free subnet(not aws subnet, bu ipv4) in vpc
+func (vpcInfo *VPCInfo) FindBiggestFreeSubnet() string {
+	var arr AddrSlice
+	// add min and max for vpc
+	net, _ := address.ParseCIDR(vpcInfo.CIDR)
+	arr = append(arr, net.Start())
+	arr = append(arr, net.End()-1)
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloud.go]: I can find the docs, but it is usually the .2 address in the vpc ...  Darn aws :)  I may be completely off btw

File: upup/pkg/fi/cloud.go [hunk 2]
```diff
@@ -169,3 +173,41 @@ func GuessCloudForZone(zone string) (CloudProviderID, bool) {
 	c, found := zonesToCloud[zone]
 	return c, found
 }
+
+// for sorting ip addresses
+type AddrSlice []address.Address
+
+func (s AddrSlice) Len() int           { return len(s) }
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloud.go]: Do these `func` need to be public?

File: upup/pkg/fi/cloud.go [hunk 3]
```diff
@@ -169,3 +173,41 @@ func GuessCloudForZone(zone string) (CloudProviderID, bool) {
...
+	arr = append(arr, net.End()-1)
+	// add min and max for subnets
+	for _, subnet := range vpcInfo.Subnets {
+		net, _ := address.ParseCIDR(subnet.CIDR)
+		arr = append(arr, net.Start())
+		arr = append(arr, net.End()-1)
+	}
+	// sort for group by pair [vpcStart, subnetStart, subnetEnd, vpcEnd]
+	sort.Sort(AddrSlice(arr))
+	// so freesubnets are [vpcStart, subnetStart], [subnetEnd, subnetXStart], [subnetYEnd, vpcEnd]
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloud.go]: This comments are awesome.  Can we get a unit test showing this?

File: upup/pkg/fi/cloud.go [hunk 4]
```diff
@@ -169,3 +173,41 @@ func GuessCloudForZone(zone string) (CloudProviderID, bool) {
...
 }
+
+// for sorting ip addresses
+type AddrSlice []address.Address
+
+func (s AddrSlice) Len() int           { return len(s) }
+func (s AddrSlice) Less(i, j int) bool { return s[i] < s[j] }
+func (s AddrSlice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// input vpc info, output biggest free subnet(not aws subnet, bu ipv4) in vpc
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloud.go]: Let me be a bit more clear :)  Please make the func comments a sentence starting with the name of the func.

File: upup/pkg/fi/cloud.go [hunk 5]
```diff
@@ -16,7 +16,11 @@ limitations under the License.
```
Comment (kenden) [inline upup/pkg/fi/cloud.go]: I think there are a lot of edge cases, so it would be good to add tests, yes!
- What if there is no ip range available for new subnets in the vpc?
- What if the leftover ip range in the vpc is so small that it's not possible to create 3 subnets in it?
- What if the 3 subnets that can be created are so small that it's not worth creating? (Ex: /32 range...)
- What if the existing subnets leave some free Ip ranges only in between them...)
- Etc

**Comment by pronix** [comment]
@chrislovecnm this one https://github.com/weaveworks/weave/blob/master/net/address/address.go

**Comment by chrislovecnm** [comment]
Can we refactor that code, see if you can remove common package, reference where we got it, and copy the file in?

**Comment by pronix** [comment]
@chrislovecnm added weave as deps in independent PR https://github.com/kubernetes/kops/pull/2400

File: upup/pkg/fi/cloud.go (hunk 1 above)
Comment (pronix) [inline upup/pkg/fi/cloud.go]: why we need care about dns and other reserved ip ? we just take big vpc subnet, subtract existed subnet and use biggest free subnet for k8s. all reserved ip addresses inside subnet is problem of aws. or i something missed ?

File: upup/pkg/fi/cloud.go (hunk 2 above)
Comment (pronix) [inline upup/pkg/fi/cloud.go]: if i correct understand https://golang.org/src/sort/sort.go - yes.

File: upup/pkg/fi/cloud.go (hunk 3 above)
Comment (pronix) [inline upup/pkg/fi/cloud.go]: unit test added

File: upup/pkg/fi/cloud.go (hunk 4 above)
Comment (pronix) [inline upup/pkg/fi/cloud.go]: thanks. changed. i pray to `go fmt` and `go vet` :)

**Comment by justinsb** [comment]
@pronix weave is a really big dependency to take.  If it's just for some CIDR math (is that right?), I'd rather not do that.

**Comment by chrislovecnm** [comment]
@pronix to be clear, we love the code that weave has - we would like to re-use the code, just not vendor all of weave code base. I attempted to communicate that here: https://github.com/kubernetes/kops/pull/2395#issuecomment-296111297  @justinsb agreed?

**Comment by pronix** [comment]
@justinsb yes. it is only for ipv4 math.
@chrislovecnm i will try to reduce dependencies.

**Comment by blakebarnett** [comment]
@pronix are you still hoping to get this into the next release?

**Comment by pronix** [comment]
@blakebarnett sorry, no time for that. may will next year.

**Comment by NullVoxPopuli** [comment]
any updates?

**Comment by dictvm** [comment]
This is still an ongoing issue. Is anyone working on this? Thanks!
